### PR TITLE
Implementing PatchTS model

### LINKED ISSUE #1277: Implement PatchTS model
### 🚀 Feature Request

Implement model [PatchTS](https://github.com/yuqinie98/PatchTST/tree/main/PatchTST_supervised)

### Proposal

Add new model to models

### Test cases

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by malodetz**
## Before submitting (must do checklist)

- [x] Did you read the [contribution guide](https://github.com/tinkoff-ai/etna/blob/master/CONTRIBUTING.md)?
- [x] Did you update the docs? We use Numpy format for all the methods and classes.
- [x] Did you write any new necessary tests?
- [x] Did you update the [CHANGELOG](https://github.com/tinkoff-ai/etna/blob/master/CHANGELOG.md)?

## Proposed Changes

See #1277 

## Closing issues

Closes #1277

### CODE REVIEW

File: etna/models/nn/patchts.py [hunk 1]
```diff
@@ -0,0 +1,379 @@
...
+            parameters for optimizer for Adam optimizer (api reference :py:class:`torch.optim.Adam`)
+        """
+        super().__init__()
+        self.patch_len = patch_len
+        self.num_layers = num_layers
+        self.feedforward_size = feedforward_size
+        self.hidden_size = hidden_size
+        self.nhead = nhead
+        self.stride = stride
+        self.loss = torch.nn.MSELoss() if loss is None else loss
```
Comment (martins0n) [inline etna/models/nn/patchts.py:106]: it's better to explicit `self.loss = loss` here because we pass this argument upstream

File: etna/models/nn/patchts.py [hunk 2]
```diff
@@ -0,0 +1,379 @@
...
+            outputs.append(pred)
+            x = torch.cat((x[:, 1:, :], torch.unsqueeze(decoder_real[:, i, :], dim=1)), dim=1)
+
+        target_prediction = torch.cat(outputs, dim=1)
+        target_prediction = torch.unsqueeze(target_prediction, dim=2)
+
+        loss = self.loss(target_prediction, decoder_target)
+        return loss, decoder_target, target_prediction
+
+    def make_samples(self, df: pd.DataFrame, encoder_length: int, decoder_length: int) -> Iterator[dict]:
```
Comment (martins0n) [inline etna/models/nn/patchts.py:194]: we should add test cases for `make_samples` for logic checks and for regression testing

File: etna/models/nn/patchts.py [hunk 3]
```diff
@@ -0,0 +1,379 @@
...
+                "decoder_target": list(),
+                "segment": None,
+            }
+            total_length = len(values_target)
+            total_sample_length = encoder_length + decoder_length
+
+            if total_sample_length + start_idx > total_length:
+                return None
+
+            # Get shifted target and concatenate it with real values features
```
Comment (martins0n) [inline etna/models/nn/patchts.py:228]: Do we really need shifts?

File: etna/models/nn/patchts.py [hunk 4]
```diff
@@ -0,0 +1,379 @@
...
+        """Make samples from segment DataFrame."""
+        values_real = (
+            df.select_dtypes(include=[np.number])
+            .assign(target_shifted=df["target"].shift(1))
+            .drop(["target"], axis=1)
+            .pipe(lambda x: x[["target_shifted"] + [i for i in x.columns if i != "target_shifted"]])
+            .values
+        )
+        values_target = df["target"].values
+        segment = df["segment"].values[0]
```
Comment (martins0n) [inline etna/models/nn/patchts.py:204]: If original implementation not using segment like features let's drop it

File: etna/models/nn/patchts.py [hunk 5]
```diff
@@ -0,0 +1,379 @@
...
+            encoder_length: int,
+            patch_len: int = 4,
+            stride: int = 1,
+            num_layers: int = 3,
+            hidden_size: int = 128,
+            feedforward_size: int = 256,
+            nhead: int = 16,
+            lr: float = 1e-3,
+            loss: Optional["torch.nn.Module"] = None,
+            train_batch_size: int = 16,
```
Comment (martins0n) [inline etna/models/nn/patchts.py:279]: If it's ok to make batch larger - we wouldn't have problems with memory, let's make  it

File: etna/models/nn/patchts.py [hunk 6]
```diff
@@ -0,0 +1,379 @@
...
+        Returns
+        -------
+        :
+            Grid to tune.
+        """
+        return {
+            "num_layers": IntDistribution(low=1, high=3),
+            "hidden_size": IntDistribution(low=4, high=64, step=4),
+            "lr": FloatDistribution(low=1e-5, high=1e-2, log=True),
+            "encoder_length": IntDistribution(low=1, high=20),
```
Comment (martins0n) [inline etna/models/nn/patchts.py:378]: Seems patch length  depends on encoder_length. Will it work for low=1?

**Review by martins0n — CHANGES_REQUESTED** [review]

**Review by martins0n — COMMENTED** [review]
Can we add examples to notebook?

File: etna/models/nn/patchts.py (hunk 4 above)
Comment (malodetz) [reply etna/models/nn/patchts.py:200]: It's needed by `raw_predict` method of `DeepBaseModel`

File: etna/models/nn/patchts.py [hunk 7]
```diff
@@ -0,0 +1,369 @@
...
+        ----------
+        x:
+            batch of data
+
+        Returns
+        -------
+        :
+            forecast with shape (batch_size, decoder_length, 1)
+        """
+        encoder_real = x["encoder_real"].float()  # (batch_size, encoder_length-1, input_size)
```
Comment (martins0n) [inline etna/models/nn/patchts.py:129]: Is it really `encoder_length-1`?

File: examples/NN_examples.ipynb
Comment (martins0n) [inline examples/NN_examples.ipynb:1]: It's better to run all from the start ot the end

**Review by martins0n — CHANGES_REQUESTED** [review]
Fix linter errors please

**Review by martins0n — APPROVED** [review]

**Review by martins0n — APPROVED** [review]
👍
